### PR TITLE
chunkserver: stat chunk bytes related  metric include WAL segment 

### PR DESCRIPTION
**Pull request by Wine93**
### What problem does this PR solve?

Problem Summary: Stat chunk bytes related metric not include WAL segment 

### What is changed and how it works?

What's Changed: Browse commit for detail

How it Works: Browse commit for detail

Side effects(Breaking backward compatibility? Performance regression?): none

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/chunkserver/chunkserver.cpp [hunk 1]
```diff
@@ -221,6 +221,11 @@ int ChunkServer::Run(int argc, char** argv) {
     copysetNodeOptions.chunkFilePool = chunkfilePool;
     copysetNodeOptions.localFileSystem = fs;
     copysetNodeOptions.trash = trash_;
+    if (kWalFilePool != nullptr) {
+        FilePoolState_t currentStat = chunkFilePoolPtr_->GetState();
+        uint32_t maxWalSegmentSize = currentStat.chunkSize + currentStat.metaPageSize;
```
Comment (SeanHai) [inline src/chunkserver/chunkserver.cpp]: maxWalSegmentSize shuold be maxWalChunkSize ?

File: src/chunkserver/config_info.h [hunk 2]
```diff
@@ -86,6 +86,8 @@ struct CopysetNodeOptions {
     uint32_t port;
     // chunk文件的大小
     uint32_t maxChunkSize;
+    // WAL segment file size
+    uint32_t maxWalSegmentSize;
```
Comment (SeanHai) [inline src/chunkserver/config_info.h:90]: should be maxWalChunkSize chunksize+metaPageSize ?

File: src/chunkserver/chunkserver.cpp [hunk 3]
```diff
@@ -152,12 +151,12 @@ int ChunkServer::Run(int argc, char** argv) {
         if (!useChunkFilePoolAsWalPool) {
             FilePoolOptions walFilePoolOptions;
             InitWalFilePoolOptions(&conf, &walFilePoolOptions);
-            kWalFilePool = std::make_shared<FilePool>(fs);
-            LOG_IF(FATAL, false == kWalFilePool->Initialize(walFilePoolOptions))
+            walFilePool = std::make_shared<FilePool>(fs);
```
Comment (SeanHai) [inline src/chunkserver/chunkserver.cpp:154]: What‘s the problem with global variables in this scenario？

File: src/chunkserver/chunkserver.cpp [hunk 4]
```diff
@@ -221,9 +221,9 @@ int ChunkServer::Run(int argc, char** argv) {
     copysetNodeOptions.walFilePool = walFilePool;
     copysetNodeOptions.localFileSystem = fs;
     copysetNodeOptions.trash = trash_;
-    if (kWalFilePool != nullptr) {
-        FilePoolState_t currentStat = chunkFilePoolPtr_->GetState();
-        uint32_t maxWalSegmentSize = currentStat.chunkSize + currentStat.metaPageSize;
+    if (nullptr != walFilePool) {
+        FilePoolOptions poolOpt = walFilePool->GetFilePoolOpt();
+        uint32_t maxWalSegmentSize = poolOpt.fileSize + poolOpt.metaPageSize;
```
Comment (SeanHai) [inline src/chunkserver/chunkserver.cpp:226]: the same problem, maxWalSegmentSize should be chunksize ?

File: src/chunkserver/config_info.h (hunk 2 above)
Comment (Wine93) [reply src/chunkserver/config_info.h:90]: YES

File: src/chunkserver/trash.cpp [hunk 5]
```diff
@@ -303,25 +306,30 @@ bool Trash::IsWALFile(const std::string &fileName) {
...
+    };
+
+    return count(copysetPath + "/" + RAFT_DATA_DIR)
+        + count(copysetPath + "/" + RAFT_LOG_DIR);
 }
 
 }  // namespace chunkserver
 }  // namespace curve
 
+
```
[the remark this answers is not in the document]
Comment (Wine93) [reply src/chunkserver/trash.cpp:335]: done

File: test/chunkserver/trash_test.cpp [hunk 6]
```diff
@@ -608,57 +623,81 @@ TEST_F(TrashTest, test_chunk_num_statistic) {
...
         .WillOnce(DoAll(SaveArg<1>(&trashedCopysetDir), Return(0)));
     std::vector<std::string> chunks3{"chunk_333", "chunk_444"};
+    std::vector<std::string> logfiles3{"curve_log_10086_10087"};
     EXPECT_CALL(*lfs, List(_, _))
-        .WillOnce(DoAll(SetArgPointee<1>(chunks3), Return(0)));
+        .WillOnce(DoAll(SetArgPointee<1>(chunks3), Return(0)))
+        .WillOnce(DoAll(SetArgPointee<1>(logfiles3), Return(0)));
     ASSERT_EQ(0, trash->RecycleCopySet(copysetDir));
-    ASSERT_EQ(7, trash->GetChunkNum());
+    ASSERT_EQ(8, trash->GetChunkNum());
```
[the remark this answers is not in the document]
Comment (Wine93) [reply test/chunkserver/trash_test.cpp:631]: @SeanHai 
In the original design, we will ignore the old format log, like `log_10083_10084 ` and `log_inprogress_10085 `, so the result is `8`, they are:
`chunk_100 `, `chunk_101 `, `chunk_200_snap_1 `, `curve_log_10086_10087 `, `curve_log_inprogress_10088 `, `chunk_333 `, `chunk_444 `, `curve_log_10086_10087 `.

But I think it's better to count the old format log for compatibility.

File: src/chunkserver/chunkserver.cpp (hunk 3 above)
Comment (Wine93) [reply src/chunkserver/chunkserver.cpp:154]: @SeanHai 

There are several reasons:
* Global variable is inelegant 
* When we stat the total number of WAL file, we need aggregate the number of WAL file in ervery copyset, so we need to know which copyset each `log storage` belongs to
* [See also](https://github.com/opencurve/curve/pull/309/files#diff-8a7f5ab58978d7cb1a50c7d167d34cacbcaeff88f246cc2117296285eb4eba23R191) for detail

File: src/chunkserver/chunkserver.cpp (hunk 4 above)
Comment (Wine93) [reply src/chunkserver/chunkserver.cpp:226]: ditto

File: src/chunkserver/chunkserver.cpp (hunk 1 above)
Comment (Wine93) [reply src/chunkserver/chunkserver.cpp]: @SeanHai 
In chunkserver contxt, we prefer to call it segment which used for WAL. :)

**Comment by Wine93** [comment]
recheck

File: src/chunkserver/heartbeat.cpp [hunk 7]
```diff
@@ -238,12 +238,19 @@ int Heartbeat::BuildRequest(HeartbeatRequest* req) {
...
+                                * walSegmentFileSize;
     uint64_t trashedChunkSize = metric->GetChunkTrashedCount() * chunkFileSize;
     uint64_t leftChunkSize = metric->GetChunkLeftCount() * chunkFileSize;
-    stats->set_chunksizeusedbytes(usedChunkSize);
-    stats->set_chunksizeleftbytes(leftChunkSize);
+    // leftWalSegmentSize will be 0 when CHUNK and WAL share file pool
+    uint64_t leftWalSegmentSize = metric->GetWalSegmentLeftCount()
+                                * walSegmentFileSize;
+
+    stats->set_chunksizeusedbytes(usedChunkSize+usedWalSegmentSize);
```
Comment (xu-chaojie) [inline src/chunkserver/heartbeat.cpp:252]: 'chunk used bytes' and 'wal segment used bytes' should be divided into two  fields in heartbeat.

File: src/chunkserver/heartbeat.cpp [hunk 8]
```diff
@@ -238,12 +238,19 @@ int Heartbeat::BuildRequest(HeartbeatRequest* req) {
...
+    stats->set_chunksizeleftbytes(leftChunkSize+leftWalSegmentSize);
```
Comment (xu-chaojie) [inline src/chunkserver/heartbeat.cpp:253]: 'left chunk  bytes' and 'left wal bytes' should be divided into two fields in heartbeat.

File: src/chunkserver/trash.cpp [hunk 9]
```diff
@@ -303,25 +306,28 @@ bool Trash::IsWALFile(const std::string &fileName) {
...
-                         << ", filename: " << chunk;
-            continue;
+    auto count = [&](const std::string& path) {
+        std::vector<std::string> chunks;
+        localFileSystem_->List(path, &chunks);
+
+        uint32_t chunkNum = 0;
+        for (auto& chunk : chunks) {
+            // valid: chunkfile, snapshotfile, walfile
+            if (!(IsChunkOrSnapShotFile(chunk) || IsWALFile(chunk))) {
```
Comment (xu-chaojie) [inline src/chunkserver/trash.cpp:316]: if we both use walpool and chunkfilepool, the 'wal in trash' and the 'chunk in trash'  should be two field.

**Comment by Wine93** [comment]
recheck

File: src/chunkserver/raftlog/curve_segment_log_storage.cpp [hunk 10]
```diff
@@ -710,5 +733,11 @@ scoped_refptr<Segment> CurveSegmentLogStorage::open_segment(
     return _open_segment;
 }
 
+LogStorageStatus CurveSegmentLogStorage::GetStatus() {
+    uint32_t count = (uint32_t)(_segments.size())
+                   + (nullptr != _open_segment ? 1 : 0);
+    return LogStorageStatus(count);
```
Comment (xu-chaojie) [inline src/chunkserver/raftlog/curve_segment_log_storage.cpp:739]: the wal segment used is not just here in the _segments,   see pop_segments().

**Comment by Wine93** [comment]
recheck

**Review by Wine93 — COMMENTED** [review]
LGTM!

File: src/chunkserver/heartbeat.cpp (hunk 7 above)
Comment (Wine93) [reply src/chunkserver/heartbeat.cpp:252]: @xu-chaojie 
This optimization will be in the another PR.

File: src/chunkserver/heartbeat.cpp (hunk 8 above)
Comment (Wine93) [reply src/chunkserver/heartbeat.cpp:253]: ditto

File: src/chunkserver/trash.cpp (hunk 9 above)
Comment (Wine93) [reply src/chunkserver/trash.cpp:316]: ditto

**Review by xu-chaojie — APPROVED** [review]

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Review by SeanHai — APPROVED** [review]

**Pull request merged**